integration/networking/bridge_test.go

```diff
@@ -0,0 +1,350 @@
+// 1. DNS resolution ; 2. ARP/NDP ; 3. whether containers can communicate with each other.
+func TestBridgeICC(t *testing.T) {
+	skip.If(t, testEnv.DaemonInfo.OSType == "windows")
+
+	d := daemon.New(t)
+	d.StartWithBusybox(t, "-D", "--experimental", "--ip6tables")
+	defer d.Stop(t)
+
+	c := d.NewClientT(t)
+	defer c.Close()
+
+	testcases := []struct {
+		name       string
+		pingCmd    []string
+		bridgeOpts []func(create *types.NetworkCreate)
+		linkLocal  bool
+		skipMsg    string
+	}{
+		{
+			name:       "IPv4 non-internal network",
+			pingCmd:    []string{"ping", "-c1", "-W3"},
+			bridgeOpts: []func(create *types.NetworkCreate){},
+		},
+		{
+			name:    "IPv4 internal network",
+			pingCmd: []string{"ping", "-c1", "-W3"},
+			bridgeOpts: []func(create *types.NetworkCreate){
+				network.WithInternal(),
+			},
+		},
+		{
+			name:    "IPv6 ULA on non-internal network",
+			pingCmd: []string{"ping", "-c1", "-W3", "-6"},
+			bridgeOpts: []func(create *types.NetworkCreate){
+				network.WithIPv6(),
+				network.WithIPAM("fdf1:a844:380c:b200::/64", "fdf1:a844:380c:b200::1"),
+			},
+		},
+		{
+			name:    "IPv6 ULA on internal network",
+			pingCmd: []string{"ping", "-c1", "-W3", "-6"},
+			bridgeOpts: []func(create *types.NetworkCreate){
+				network.WithIPv6(),
+				network.WithInternal(),
+				network.WithIPAM("fdf1:a844:380c:b247::/64", "fdf1:a844:380c:b247::1"),
+			},
+			skipMsg: "See moby/moby#45649",
+		},
+		{
+			name:    "IPv6 link-local address on non-internal network",
+			pingCmd: []string{"ping", "-c1", "-W3", "-6"},
+			bridgeOpts: []func(create *types.NetworkCreate){
+				network.WithIPv6(),
+				// There's no real way to specify an IPv6 network is only used with SLAAC link-local IPv6 addresses.
+				// What we can do instead, is to tell the IPAM driver to assign addresses from the link-local prefix.
+				// Each container will have two link-local addresses: 1. a SLAAC address assigned by the kernel ;
+				// 2. the one dynamically assigned by the IPAM driver.
+				network.WithIPAM("fe80::/64", "fe80::1"),
+			},
+			linkLocal: true,
+		},
+		{
+			name:    "IPv6 link-local address on internal network",
+			pingCmd: []string{"ping", "-c1", "-W3", "-6"},
+			bridgeOpts: []func(create *types.NetworkCreate){
+				network.WithIPv6(),
+				network.WithInternal(),
+				// See the note above about link-local addresses.
+				network.WithIPAM("fe80::/64", "fe80::1"),
+			},
+			linkLocal: true,
+			skipMsg:   "See moby/moby#45649",
+		},
+	}
+
+	for tcID, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			skip.If(t, tc.skipMsg != "", tc.skipMsg)
+			ctx := context.Background()
+
+			bridgeName := fmt.Sprintf("testnet-icc-%d", tcID)
+			network.CreateNoError(ctx, t, c, bridgeName, append(tc.bridgeOpts,
+				network.WithDriver("bridge"),
+				network.WithOption("com.docker.network.bridge.name", bridgeName))...)
+			defer network.RemoveNoError(ctx, t, c, bridgeName)
```

Comment:
Should these use `t.Cleanup()` instead of `defer` ? Or was there a specific reason to use `defer` ?